### PR TITLE
Added worker types

### PR DESCRIPTION
**Pull request by swarnim-vyas**
This change adds worker types on how many threads they have and if they are GPU enabled or not.
Also updated relevant proto files and types to ensure that register worker passes all this information to scheduler and scheduler is correctly able to parse it over grpc protocol.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,10 +14,14 @@ Steps to Run
 4) Run compile_proto script based on your target machine from root repo folder (CS244B)
 5) Start the scheduler server from one terminal - python3 scheduler_server.py
 6) Start the worker server in another terminal - python3 worker_server.py <PortNumber>
-    Example Command: python3 worker_server.py 50054
-    Example Command: python3 worker_server.py 50054 --addDelay
+    Example Command: python3 worker_server.py 50054 5
+    Example Command: python3 worker_server.py 50053 --addDelay 5 --gpuEnabled
```
Comment (krish8484) [inline README.md]: MaxWorkerCount missing from this example?

File: worker_server.py [hunk 2]
```diff
@@ -41,19 +45,29 @@ def serve():
...
 if __name__ == '__main__':
-    parser = argparse.ArgumentParser(description = 'Please pass the port number and DelayFlag')
+    parser = argparse.ArgumentParser()
     parser.add_argument('PortNumber', help='Please pass the portNumber on which you want the worker to listen.')
     
     parser.add_argument("--addDelay", action="store_true", 
-                    help="Please pass the AddDelay parameter. This adds a random delay in worker operations.") 
+                    help="Please pass the AddDelay parameter. This adds a random delay in worker operations.")
+    
+    parser.add_argument('MaxWorkerCount', help='Please pass the max thread that you want worker to spawn.')
```
Comment (krish8484) [inline worker_server.py]: keep this as optional and set a default?

File: scheduler.py [hunk 3]
```diff
@@ -38,19 +39,18 @@ def submit_task(self, task : Task) -> Future:
             else:
                 logging.error(f"SubmitTask: Unhandled RPC error: code={rpc_error.code()} message={rpc_error.details()}")
 
-    def register_worker(self, host_name, port_number):
+    def register_worker(self, workerInfo):
+        print("In register worker at scheduler")
```
Comment (krish8484) [inline scheduler.py:43]: remove debug print

File: scheduler_server.py [hunk 4]
```diff
@@ -27,7 +28,13 @@ def TaskCompleted(self, request, context):
             return api_pb2.TaskCompletedResponse(status = api_pb2.Status(success = False))
     
     def RegisterWorker(self, request, context):
-        assignedWorkerId = self.scheduler.register_worker(request.hostName, request.portNumber)
+        print("In scheduler server")
```
Comment (krish8484) [inline scheduler_server.py]: remove debug print

File: worker_server.py [hunk 5]
```diff
@@ -13,6 +13,8 @@
 WORKER_HOST = "localhost"
 WORKER_PORT = 50052
 ADD_DELAY = False
+MAX_WORKERS = 2
```
Comment (krish8484) [inline worker_server.py]: can we rename to max_thread_per_worker. MAX_WORKERS seems like a misnomer? Similar for other variables used for this purpose in the file

File: README.md (hunk 1 above)
Comment (krish8484) [reply README.md]: it seems like the instructions below and the command do not match

Comment (swarnim-vyas) [reply README.md]: The '5' is the worker count

File: worker_server.py (hunk 2 above)
Comment (swarnim-vyas) [reply worker_server.py]: Kept it mandatory so that we trigger the worker knowing its capacity

File: scheduler.py (hunk 3 above)
Comment (swarnim-vyas) [reply scheduler.py:43]: removed

File: worker_server.py (hunk 5 above)
Comment (swarnim-vyas) [reply worker_server.py]: Updated

File: Protos/api.proto [hunk 6]
```diff
@@ -66,6 +65,13 @@ message Status {
     bool success = 1;
 }
 
+
+message WorkerInfo {
+    string hostName = 1;
+    int32 portNumber = 2;
+    int32 maxThreadCount = 3;
+    bool isGPUEnabled = 4;
+
```
Comment (aadhithNarayan) [inline Protos/api.proto:75]: I think CI is failing due to unclosed brackets. Please take a look into this.
@swarnim-vyas did you get a chance to test this out?

File: README.md [hunk 7]
```diff
@@ -14,10 +14,14 @@ Steps to Run
...
+
+    First parameter is the port number
+    Second parameter is the number of threads the worker can process
+    Third parameter is an optional parameter. If --addDelay parameter is passed, delays are added randomly for the worker operations.
```
Comment (aadhithNarayan) [inline README.md]: The order is different from the code: 2 <--> 3 or there is no order?

File: worker_server.py (hunk 2 above)
Comment (aadhithNarayan) [inline worker_server.py]: Can we rename MaxWorkerCount since it seems ambiguous with Worker context that we have?

File: worker_server.py [hunk 8]
```diff
@@ -32,7 +36,7 @@ def SubmitTask(self, request, context):
         return api_pb2.TaskResponse(future=api_pb2.Future(resultLocation=future.resultLocation, hostName=future.hostName, port=future.port))
 
 def serve():
-    server = grpc.server(futures.ThreadPoolExecutor())
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers = MAX_WORKERS_COUNT))
```
Comment (aadhithNarayan) [inline worker_server.py]: Threadpool here is for gRPC to accept connections concurrently in multiple threads. I think a more appropriate constraint should be done in `submit_task` method in worker.py where we execute the task in a thread (spawning of threads here should be capped based on user input)

File: scheduler_client.py [hunk 9]
```diff
@@ -3,6 +3,8 @@
 import grpc
 from Data.future import Future
 from Data.task import Task
+from Data.WorkerInfo import WorkerInfo
+from Data.WorkerInfo import WorkerInfo
```
Comment (aadhithNarayan) [inline scheduler_client.py]: Please remove redundant imports

File: Protos/api.proto (hunk 6 above)
Comment (swarnim-vyas) [reply Protos/api.proto:75]: Yeah - tested it out but I think manual merge with main caused all these issues. 
Fixing them.

File: worker_server.py (hunk 8 above)
Comment (swarnim-vyas) [reply worker_server.py]: I would like to discuss this more - lets get this PR in and we can then modify after our discussion tomorrow.

File: worker_server.py (hunk 2 above)
Comment (swarnim-vyas) [reply worker_server.py]: Updated

File: README.md (hunk 7 above)
Comment (swarnim-vyas) [reply README.md]: Updated the usage

File: Protos/api.proto (hunk 6 above)
Comment (swarnim-vyas) [reply Protos/api.proto:75]: Updated - please take a look.

File: worker_server.py (hunk 8 above)
Comment (swarnim-vyas) [reply worker_server.py]: Updated

File: README.md (hunk 1 above)
Comment (aadhithNarayan) [inline README.md]: Shouldn't 5 come before --addDelay?

**Comment by krish8484** [comment]
Would be nice to have a test for this feature.

**Review by krish8484 — APPROVED** [review]

**Review by aadhithNarayan — APPROVED** [review]

**Pull request merged**